File: src/co/aospa/sense/camera/CameraService.java
package co.aospa.sense.camera;

import android.graphics.SurfaceTexture;
import android.hardware.Camera;
import android.os.Handler;
import android.view.SurfaceHolder;

import co.aospa.sense.camera.callables.AddCallbackBufferCallable;
import co.aospa.sense.camera.callables.AutoFocusCallable;
import co.aospa.sense.camera.callables.CameraCallable;
import co.aospa.sense.camera.callables.CloseCameraCallable;
import co.aospa.sense.camera.callables.OpenCameraCallable;
import co.aospa.sense.camera.callables.ReadParamsCallable;
import co.aospa.sense.camera.callables.SetDisplayOrientationCallback;
import co.aospa.sense.camera.callables.SetFaceDetectionCallback;
import co.aospa.sense.camera.callables.SetPreviewCallbackCallable;
import co.aospa.sense.camera.callables.StartPreviewCallable;
import co.aospa.sense.camera.callables.StopPreviewCallable;
import co.aospa.sense.camera.callables.WriteParamsCallable;
import co.aospa.sense.camera.listeners.ByteBufferCallbackListener;
import co.aospa.sense.camera.listeners.CameraListener;
import co.aospa.sense.camera.listeners.ErrorCallbackListener;
import co.aospa.sense.camera.listeners.FocusResultListener;
import co.aospa.sense.camera.listeners.ReadParametersListener;

public class CameraService {

    private static final int DEFAULT_MSG_TYPE = 1;
    private final Handler mServiceHandler;

    private CameraService() {
        CameraHandlerThread cameraHandlerThread = new CameraHandlerThread();
        cameraHandlerThread.start();
        this.mServiceHandler = new Handler(cameraHandlerThread.getLooper(), message -> {
            ((CameraCallable) message.obj).run();
            return true;
        });
    }

    private static CameraService getInstance() {
        return LazyLoader.INSTANCE;
    }

    public static void openCamera(int id, ErrorCallbackListener errorCallbackListener, CameraListener cameraListener) {
        getInstance().addCallable(new OpenCameraCallable(id, errorCallbackListener, cameraListener));
    }

    public static void closeCamera(CameraListener cameraListener) {
        CameraService instance = getInstance();
        clearQueue();
        instance.addCallable(new CloseCameraCallable(cameraListener));
    }

    public static void autoFocus(boolean enable, FocusResultListener focusResultListener, CameraListener cameraListener) {
        getInstance().addCallable(new AutoFocusCallable(enable, focusResultListener, cameraListener));
    }

    public static void readParameters(ReadParametersListener readParametersListener, CameraListener cameraListener) {
        getInstance().addCallable(new ReadParamsCallable(readParametersListener, cameraListener));
    }

    public static void writeParameters(CameraListener cameraListener) {
        getInstance().addCallable(new WriteParamsCallable(cameraListener));
    }

    public static void startPreview(CameraListener cameraListener) {
        getInstance().addCallable(new StartPreviewCallable(cameraListener));
    }

    public static void startPreview(SurfaceTexture surfaceTexture, CameraListener cameraListener) {
        getInstance().addCallable(new StartPreviewCallable(surfaceTexture, cameraListener));
    }

    public static void startPreview(SurfaceHolder surfaceHolder, CameraListener cameraListener) {
        getInstance().addCallable(new StartPreviewCallable(surfaceHolder, cameraListener));
    }

    public static void stopPreview(CameraListener cameraListener) {
        getInstance().addCallable(new StopPreviewCallable(cameraListener));
    }

    public static void addCallbackBuffer(byte[] data, CameraListener cameraListener) {
        getInstance().addCallable(new AddCallbackBufferCallable(data, cameraListener));
    }

    public static void setPreviewCallback(ByteBufferCallbackListener byteBufferCallbackListener, boolean withBuffer, CameraListener cameraListener) {
        getInstance().addCallable(new SetPreviewCallbackCallable(byteBufferCallbackListener, withBuffer, cameraListener));
    }

    public static void setFaceDetectionCallback(Camera.FaceDetectionListener faceDetectionListener, CameraListener cameraListener) {
        getInstance().addCallable(new SetFaceDetectionCallback(faceDetectionListener, cameraListener));
    }

    public static void setDisplayOrientationCallback(int angle, CameraListener cameraListener) {
        getInstance().addCallable(new SetDisplayOrientationCallback(angle, cameraListener));
    }

    public static void clearQueue() {
        getInstance().mServiceHandler.removeMessages(DEFAULT_MSG_TYPE);
    }

    private void addCallable(CameraCallable cameraCallable) {
        this.mServiceHandler.sendMessage(this.mServiceHandler.obtainMessage(DEFAULT_MSG_TYPE, cameraCallable));
    }

    private static final class LazyLoader {
        private static final CameraService INSTANCE = new CameraService();

        private LazyLoader() {
        }
    }
}

File: src/co/aospa/sense/camera/CameraService.kt
package co.aospa.sense.camera

import android.graphics.SurfaceTexture
import android.hardware.Camera
import android.os.Handler
import android.os.Message
import android.view.SurfaceHolder
import co.aospa.sense.camera.callables.CameraCallable
import co.aospa.sense.camera.listeners.CameraEventListener
import co.aospa.sense.camera.listeners.CameraListener
import co.aospa.sense.camera.callables.OpenCameraCallable
import co.aospa.sense.camera.callables.CloseCameraCallable
import co.aospa.sense.camera.callables.ReadParamsCallable
import co.aospa.sense.camera.callables.WriteParamsCallable
import co.aospa.sense.camera.callables.StartPreviewCallable
import co.aospa.sense.camera.callables.StopPreviewCallable
import co.aospa.sense.camera.callables.AddCallbackBufferCallable
import co.aospa.sense.camera.callables.SetPreviewCallbackCallable
import co.aospa.sense.camera.callables.SetFaceDetectionCallback
import co.aospa.sense.camera.callables.SetDisplayOrientationCallback

class CameraService private constructor() {

    private val mServiceHandler: Handler
    private fun addCallable(cameraCallable: CameraCallable) {
        mServiceHandler.sendMessage(mServiceHandler.obtainMessage(DEFAULT_MSG_TYPE, cameraCallable))
    }

    private object LazyLoader {
        var field: CameraService = CameraService()
    }

    companion object {
        private const val DEFAULT_MSG_TYPE = 1
        fun openCamera(id: Int, errorListener: CameraEventListener?, listener: CameraListener?) {
            val instance: CameraService = LazyLoader.field
            instance.addCallable(OpenCameraCallable(id, errorListener, listener))
        }

        fun closeCamera(listener: CameraListener?) {
            val instance: CameraService = LazyLoader.field
            clearQueue()
            instance.addCallable(CloseCameraCallable(listener))
        }

        fun readParameters(eventListener: CameraEventListener?, listener: CameraListener?) {
            val instance: CameraService = LazyLoader.field
            instance.addCallable(ReadParamsCallable(eventListener, listener))
        }

        fun writeParameters(listener: CameraListener?) {
            val instance: CameraService = LazyLoader.field
            instance.addCallable(WriteParamsCallable(listener))
        }

        fun startPreview(surfaceTexture: SurfaceTexture?, listener: CameraListener?) {
            val instance: CameraService = LazyLoader.field
            instance.addCallable(StartPreviewCallable(surfaceTexture, listener))
        }

        fun startPreview(surfaceHolder: SurfaceHolder?, listener: CameraListener?) {
            val instance: CameraService = LazyLoader.field
            instance.addCallable(StartPreviewCallable(surfaceHolder, listener))
        }

        fun stopPreview(listener: CameraListener?) {
            val instance: CameraService = LazyLoader.field
            instance.addCallable(StopPreviewCallable(listener))
        }

        fun addCallbackBuffer(data: ByteArray?, listener: CameraListener?) {
            val instance: CameraService = LazyLoader.field
            instance.addCallable(AddCallbackBufferCallable(data, listener))
        }

        fun setPreviewCallback(
            eventListener: CameraEventListener?,
            withBuffer: Boolean,
            listener: CameraListener?
        ) {
            val instance: CameraService = LazyLoader.field
            instance.addCallable(
                SetPreviewCallbackCallable(
                    eventListener,
                    withBuffer,
                    listener
                )
            )
        }

        fun setFaceDetectionCallback(
            faceDetectionListener: Camera.FaceDetectionListener?,
            listener: CameraListener?
        ) {
            val instance: CameraService = LazyLoader.field
            instance.addCallable(
                SetFaceDetectionCallback(
                    faceDetectionListener,
                    listener
                )
            )
        }

        fun setDisplayOrientationCallback(angle: Int, listener: CameraListener?) {
            val instance: CameraService = LazyLoader.field
            instance.addCallable(SetDisplayOrientationCallback(angle, listener))
        }

        fun clearQueue() {
            val instance: CameraService = LazyLoader.field
            instance.mServiceHandler.removeMessages(DEFAULT_MSG_TYPE)
        }
    }

    init {
        val cameraHandlerThread = CameraHandlerThread()
        cameraHandlerThread.start()
        mServiceHandler = Handler(cameraHandlerThread.looper) { message: Message ->
            (message.obj as CameraCallable).run()
            true
        }
    }
}